### ISSUE TITLE
Can't Download Offline Chat

### ISSUE THREAD

**Issue by werkjhwer** [irsat000/CAI-Tools#66]
Hey, I'm working off a Chromebook. Waited for the update to be ready. Error says "Unexpected CAI Tools error, please report on Github". Also tried making Tavern chat, or getting history, and cannot. Please let me know if there's anything on my end. 
Have a good one!

**Comment by irsat000** [comment]
This is a second one. Alright, probably because Character.ai changed something. I will update it.

**Comment by irsat000** [comment]
Is it fixed?

**Comment by lorenzoveee** [comment]
> Is it fixed?

Unfortunately I'm having the same issue as of ten minutes ago, so I don't think it's fixed yet

**Comment by irsat000** [comment]
I don't know how it doesn't happen to me.

Are you using the new website or the old?

**Comment by lorenzoveee** [comment]
> I don't know how it doesn't happen to me.
> 
> Are you using the new website or the old? Have you tried a different browser? Which browser are you getting this error on?

The new one. I did try Firefox (I use Chrome normally) but got the same error. It seems to only happen with characters where I have multiple chat histories, if that makes sense? Because I tried to download from a character where I only had a single chat going and it downloaded with no issue.

**Comment by binyadotorg** [comment]
Having the same issue, across two machines and browsers.
- Windows 11 PC, Microsoft Edge Version 127.0.2651.86 (Official build) (64-bit)
- Macbook Air 2020, Google Chrome Version 127.0.6533.89 (Official build) (x86_64)

Both instances are giving the error "Unexpected CAI Tools error, please report on Github" when loading now, and both are also giving me the error "Data doesn't exist or not ready. Try again later." if I try to download an offline chat/history.

**Comment by irsat000** [comment]
Alright, the reason it worked on me was because I did a couple tests. After spamming the Update button to try it again and again, also trying a character with 15 chats, I figure out the problem.

Previously the CAI servers were not giving any error when fetching chats, now apparently they do. Within expectations obviously, even if it was fine for more than a year. The real problem is the extension had no error handling for that. So it crashes if server fails to respond.

"502 Bad Gateway" this is the status code. I will fix this with a simple auto try again system, if same request fails 3 times, only then we can stop.

In summary, my code sucks, time to update.

**Comment by lorenzoveee** [comment]
> Alright, the reason it worked on me was because I did a couple tests. After spamming the Update button to try it again and again, also trying a character with 15 chats, I figure out the problem.
> 
> Previously the CAI servers were not giving any error when fetching chats, now apparently they do. Within expectations obviously, even if it was fine for more than a year. The real problem is the extension had no error handling for that. So it crashes if server fails to respond.
> 
> "502 Bad Gateway" this is the status code. I will fix this with a simple auto try again system, if same request fails 3 times, only then we can stop.
> 
> In summary, my code sucks, time to update.

Doesn't suck at all! All the work you put in is much appreciated. I'm just glad you were finally able to replicate the error, I was beginning to worry it was just me haha

**Comment by lfvjd** [comment]
I have the same issue. Loading (in the This Conversation zone) just fails after trying for a while and gives the Unexpected Error message. If I try to download the offline chat afterwards, it says that there's no data or it isn't ready. 
Tried both mac and windows and Safari, Chrome and Firefox, and it is the new website.
Character only has one chat.
Now, it is a pretty long chat. I assumed at first that it could be the reason why it's been failing...

**Comment by lfvjd** [comment]
Updating: It actually worked for me after spamming Update, as you said.
Thank you so much for your hard work!

**Comment by irsat000** [comment]
Please check the other issue thread, this is fixed. Will be in the next update.

**Comment by lorenzoveee** [comment]
> Please check the other issue thread, this is fixed. Will be in the next update.

Awesome, thank you!

**Comment by werkjhwer** [comment]
Hey, it works again!!! Thank you so much <3 And your work is amazing.

**Comment by lorenzoveee** [comment]
> I didn't update yet. Maybe cai servers are fixed.

Yeah I was going to say I'm still having the same issue, but I assumed it was cos you hadn't updated yet lol

**Comment by irsat000** [comment]
It's pretty much complete. I have some testing to do, then I can release it.

**Comment by lorenzoveee** [comment]
> It's pretty much complete. I have some testing to do, then I can release it.

Thank you for keeping us updated, your work is much appreciated

**Comment by irsat000** [comment]
Should be fixed now.
